src/mpi/coll/algorithms/treealgo/treealgo.c

```diff
@@ -62,14 +65,162 @@ int MPIR_Treealgo_tree_create(int rank, int nranks, int tree_type, int k, int ro
+                if (comm->coll.topo_aware_tree) {
+                    MPIR_Treealgo_tree_free(comm->coll.topo_aware_tree);
+                } else {
+                    comm->coll.topo_aware_tree =
+                        (MPIR_Treealgo_tree_t *) MPL_malloc(sizeof(MPIR_Treealgo_tree_t),
+                                                            MPL_MEM_BUFFER);
+                }
+                mpi_errno =
+                    MPII_Treeutil_tree_topology_aware_init(comm, k, root, enable_reorder,
+                                                           comm->coll.topo_aware_tree);
+                MPIR_ERR_CHECK(mpi_errno);
+                *ct = *comm->coll.topo_aware_tree;
+                comm->coll.topo_aware_tree_root = root;
+            }
+            *ct = *comm->coll.topo_aware_tree;
+            utarray_new(ct->children, &ut_int_icd, MPL_MEM_COLL);
+            for (int i = 0; i < ct->num_children; i++) {
+                utarray_push_back(ct->children,
+                                  &ut_int_array(comm->coll.topo_aware_tree->children)[i],
+                                  MPL_MEM_COLL);
+            }
+            break;
+
+        case MPIR_TREE_TYPE_TOPOLOGY_AWARE_K:
+            if (!comm->coll.topo_aware_k_tree || root != comm->coll.topo_aware_k_tree_root) {
+                if (comm->coll.topo_aware_k_tree) {
+                    MPIR_Treealgo_tree_free(comm->coll.topo_aware_k_tree);
+                } else {
+                    comm->coll.topo_aware_k_tree =
+                        (MPIR_Treealgo_tree_t *) MPL_malloc(sizeof(MPIR_Treealgo_tree_t),
+                                                            MPL_MEM_BUFFER);
+                }
+                mpi_errno =
+                    MPII_Treeutil_tree_topology_aware_k_init(comm, k, root, enable_reorder,
+                                                             comm->coll.topo_aware_k_tree);
+                MPIR_ERR_CHECK(mpi_errno);
+                *ct = *comm->coll.topo_aware_k_tree;
+                comm->coll.topo_aware_k_tree_root = root;
+            }
+            *ct = *comm->coll.topo_aware_k_tree;
+            utarray_new(ct->children, &ut_int_icd, MPL_MEM_COLL);
+            for (int i = 0; i < ct->num_children; i++) {
+                utarray_push_back(ct->children,
+                                  &ut_int_array(comm->coll.topo_aware_k_tree->children)[i],
+                                  MPL_MEM_COLL);
+            }
+            break;
+
+        default:
+            MPIR_ERR_SETANDJUMP1(mpi_errno, MPI_ERR_OTHER, "**treetype", "**treetype %d",
+                                 tree_type);
+    }
+
+    MPIR_FUNC_EXIT;
+
+  fn_exit:
+    if (MPIR_CVAR_TREE_DUMP) {
+        char outfile_name[PATH_MAX];
+        sprintf(outfile_name, "%s%d.json", "tree", comm->rank);
+        FILE *outfile = fopen(outfile_name, "w");
+        dump_node(ct, outfile);
+        fclose(outfile);
```

Comment:
Since we are duplicating this code a few times, maybe refactor into a static utility?

Reply:
Good idea. Created a static function "dump_tree"